man/bnbot.1
.\"
.\" Copyright (C) 2000  Ross Combs (ross@bnetd.org)
.\" 
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111,
.\" USA.
.\"
.TH BNBOT 1 "11 June, 2000" "BNETD" "BNETD User's Manual"
.SH NAME
bnbot \- Battle.net text-based chat bot client
.SH SYNOPSIS
.B bnbot
[\fIoptions\fP]
\fIservername\fP
[\fIserverport\fP]
.SH DESCRIPTION
.B bnbot
is a simple program which will allow communication through the bot protocol of a
Battle.net server from the command line.  Input from stdin is sent to the server and
output from the server is send to stdout.  The interface is so simple that
.BR telnet (1)
could be used.  The only complicating factor is that a control-C character must be sent
to the server before any other input.
.LP
It is hoped that this program can be used with a scripting language like
.BR expect (1)
to make creation of chat bots for Unix easier.  To date, I don't know of anyone doing
this.  This program may also be used to test bot login permissions for accounts.
.LP
If no \fIservername\fP is specified, localhost is assumed.  If no \fIserverport\fP is
specified, 6112 is assumed.
.SH OPTIONS
.TP
.B \-h --help --usage
Print usage information and exit.
.TP
.B \-v --version
Print its version number and exit.
.SH NOTES
The current version is available with the bnetd distribution at:
.LP
.RS
.I http://www.bnetd.org/
.RE
.SH BUGS
The terminal settings are not consulted for keybindings.  The following are assumed:
.TP
^H
delete the character to the left of the cursor
.TP
^J
accept current line
.TP
^M (return)
accept current line
.TP
^T
transpose the last two characters
.TP
^W
delete word to the left of the cursor
.TP
^U
delete the whole input line
.TP
^[ (escape)
cancel the current input line
.TP
^? (delete)
delete the character to the left of the cursor
.SH "SEE ALSO"
.BR bnetd (1),
.BR bnchat (1)
.SH AUTHOR
.TP
Ross Combs (\fIross@bnetd.org\fP)


NEWS

		    PvPGN 1.7.x Major Changes And New Features
		  ==============================================

 o Westwood Online (WOL) support

    This is a historic moment in the life of the PvPGN project. For the first 
    time a different gaming network than Battle.net is supported with PvPGN. 
    Current supports includes almost complete support for Tiberian Sun (TS) 
    and basic support for Red Alert 2 (RA2) and Yuri's Revenge (YR).

 o WIN32 GUI for d2cs and d2dbs

    d2cs and d2bs now feature a new slick GUI if build under WIN32.

 o Instant startup (partial)

    PvPGN now can startup in a lot lesser time than before, this is due to 
    code changes which make it not load all accounts at startup. This 
    improvement should be mostly noticed on very high number of accounts 
    servers. Note that currently this happens only to SQL storage types 
    (the file based storage types still load all accounts at startup).

 o Optimized account loading/saving codes

    Account loading/saving codes have been rewritten allowing PvPGN to scale 
    a lot better with a high number of online users (ie lower CPU usage).

 o Better game result calculation

    Game result decision has been completly changed to use a democratic 
    ellection scheme (ie it bases the decision on the oppinions of the most 
    results). This should solve problems with game results beeing wrong under 
    some situations (like team games).

 o New WIN32 compiler/enviroment support

    Now you can compile PvPGN under WIN32 using complete GPL/free build tools 
    by using Dev-C++ project and additional tools (mingw, etc).

 o Optimized handling of D2 clients

    Replaced a O(n) search algorithm through the client connection list used 
    mostly by D2 handling functions with a O(1) algorithm.

 o Optimized timers codes

    Timers codes have been rewritten so the most used functions complete a 
    lot faster than before. This improves PvPGN CPU usage a lot because some 
    timers functions are used very often inside PvPGN.

 o Memory management changes and optimizations

    There has been added some memory management wrappers functions which 
    never fail and thus the calling codes were simplified a lot by removing 
    checks to memory allocation failures. If memory allocation fails PvPGN 
    tries to recover by programming an immediate shutdown and cleaning up 
    some safe memory.

 o Changed codes to use logged in username instead of account username

    This change makes PvPGN use the username introduced at login (with the 
    exact case as in login) instead of the account name. This solves some 
    bugs (like could not edit profile in D2 if logged in with different case) 
    and also gets us more close to Battle.Net.

 o Switched from string based tags to integer based tags

    Codes have been dramatically changed to use integers for storing of tags 
    (client tags, architecture tags, etc) and thus when comparing tags we use 
    plain integer comparation instead of string comparation. This should 
    reduce CPU usage a lot.

 o Optimized list handling codes

    The list API has been changed and the codes have been optimized so it 
    will remove elements from the list upon request and not later, thus 
    eliminating a O(n) search and thus making PvPGN a lot faster.

 o Unified and extended address translation system

    We have unified all address translation files and codes to use a single 
    address_translation.conf file for this job. Also the file syntax has 
    been completly changed to be more flexible and allow configuration of 
    almost any thinkable NAT setup.

 o IRC support reworked

    IRC client support has been bugfixed, optimized and made to work with a 
    lot more IRC clients.

 o Unified file based storage

    File based storage (plain and cdb) have been unified into a single storage 
    backend (called "file") and the mode (plain/cdb) is just a parameter.

 o Rewritten low level socket multiplexing codes (fdwatch2)

    The codes which deal with how PvPGN knows which sockets are available for 
    reading/writing have been rewritten with a new interface which does not 
    presume (like the old codes) that the underlying OS allocates socket 
    values "close" (ex. WIN32 systems do not while most Unix based systems do 
    so). The new codes eliminates this presumtion and thus allows WIN32 hosts 
    to support a large number of connections as PvPGN supported Unix based 
    platforms.

 o Added more storage types: SQLite3 and ODBC

    With 1.7.x PvPGN now supports besides the MySQL/PostgreSQL SQL backends 
    also SQLite3 (www.sqlite.org) and ... ODBC SQL engine !! This means that 
    theoretically now you should be able to use PvPGN with almost any SQL 
    engine as long as an ODBC driver exists for it.
 

src/client/bnbot.cpp
/*
 * Copyright (C) 1999  Ross Combs (rocombs@cs.nmsu.edu)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
#include "common/setup_before.h"
#include <cstdio>
#include <cstdlib>
#include <cctype>
#include <cstring>
#include <cerrno>

#ifdef HAVE_TERMIOS_H
# include <termios.h>
#endif

#include "compat/termios.h"
#include "compat/psock.h"
#include "compat/inet_ntoa.h"
#include "common/packet.h"
#include "common/field_sizes.h"
#include "common/util.h"
#include "common/init_protocol.h"
#include "common/bn_type.h"
#include "common/network.h"
#ifdef CLIENTDEBUG
# include "common/eventlog.h"
#endif
#include "client.h"
#include "common/setup_after.h"

using namespace pvpgn;
using namespace pvpgn::client;

namespace
{

	void usage(char const * progname)
	{
		std::fprintf(stderr, "usage: %s [<options>] [<servername> [<TCP portnumber>]]\n"
			"    -h, --help, --usage         show this information and exit\n"
			"    -v, --version               print version number and exit\n",
			progname);
		std::exit(EXIT_FAILURE);
	}

}

extern int main(int argc, char * argv[])
{
	int                a;
	int                sd;
	struct sockaddr_in saddr;
	t_packet *         packet;
	t_packet *         rpacket;
	char const *       servname = NULL;
	unsigned short     servport = 0;
	char               text[MAX_MESSAGE_LEN];
	struct hostent *   host;
	unsigned int       commpos;
	struct termios     in_attr_old;
	struct termios     in_attr_new;
	int                changed_in;
	unsigned int       currsize;
	int                fd_stdin;
	unsigned int       screen_width, screen_height;

	if (argc < 1 || !argv || !argv[0])
	{
		std::fprintf(stderr, "bad arguments\n");
		return EXIT_FAILURE;
	}

	for (a = 1; a < argc; a++)
	if (servname && std::isdigit((int)argv[a][0]) && a + 1 >= argc)
	{
		if (str_to_ushort(argv[a], &servport) < 0)
		{
			std::fprintf(stderr, "%s: \"%s\" should be a positive integer\n", argv[0], argv[a]);
			usage(argv[0]);
		}
	}
	else if (!servname && argv[a][0] != '-' && a + 2 >= argc)
		servname = argv[a];
	else if (std::strcmp(argv[a], "-v") == 0 || std::strcmp(argv[a], "--version") == 0)
	{
		std::printf("version "PVPGN_VERSION"\n");
		return EXIT_SUCCESS;
	}
	else if (std::strcmp(argv[a], "-h") == 0 || std::strcmp(argv[a], "--help") == 0 || std::strcmp(argv[a], "--usage") == 0)
		usage(argv[0]);
	else
	{
		std::fprintf(stderr, "%s: unknown option \"%s\"\n", argv[0], argv[a]);
		usage(argv[0]);
	}

	if (servport == 0)
		servport = BNETD_SERV_PORT;
	if (!servname)
		servname = BNETD_DEFAULT_HOST;

	if (psock_init() < 0)
	{
		std::fprintf(stderr, "%s: could not inialialize socket functions\n", argv[0]);
		return EXIT_FAILURE;
	}

	if (!(host = gethostbyname(servname)))
	{
		std::fprintf(stderr, "%s: unknown host \"%s\"\n", argv[0], servname);
		return EXIT_FAILURE;
	}

	fd_stdin = fileno(stdin);
	if (tcgetattr(fd_stdin, &in_attr_old) >= 0)
	{
		in_attr_new = in_attr_old;
		in_attr_new.c_lflag &= ~(ECHO | ICANON); /* turn off ECHO and ICANON */
		in_attr_new.c_cc[VMIN] = 0; /* don't require reads to return data */
		in_attr_new.c_cc[VTIME] = 1; /* timeout = .1 second */
		tcsetattr(fd_stdin, TCSANOW, &in_attr_new);
		changed_in = 1;
	}
	else
	{
		std::fprintf(stderr, "%s: could not set terminal attributes for stdin\n", argv[0]);
		changed_in = 0;
	}

	if (client_get_termsize(fd_stdin, &screen_width, &screen_height) < 0)
	{
		std::fprintf(stderr, "%s: could not determine screen size\n", argv[0]);
		if (changed_in)
			tcsetattr(fd_stdin, TCSAFLUSH, &in_attr_old);
		return EXIT_FAILURE;
	}

	if ((sd = psock_socket(PSOCK_PF_INET, PSOCK_SOCK_STREAM, PSOCK_IPPROTO_TCP)) < 0)
	{
		std::fprintf(stderr, "%s: could not create socket (psock_socket: %s)\n", argv[0], std::strerror(psock_errno()));
		if (changed_in)
			tcsetattr(fd_stdin, TCSAFLUSH, &in_attr_old);
		return EXIT_FAILURE;
	}

	std::memset(&saddr, 0, sizeof(saddr));
	saddr.sin_family = PSOCK_AF_INET;
	saddr.sin_port = htons(servport);
	std::memcpy(&saddr.sin_addr.s_addr, host->h_addr_list[0], host->h_length);
	if (psock_connect(sd, (struct sockaddr *)&saddr, sizeof(saddr)) < 0)
	{
		std::fprintf(stderr, "%s: could not connect to server \"%s\" port %hu (psock_connect: %s)\n", argv[0], servname, servport, std::strerror(psock_errno()));
		if (changed_in)
			tcsetattr(fd_stdin, TCSAFLUSH, &in_attr_old);
		return EXIT_FAILURE;
	}

	if (psock_ctl(sd, PSOCK_NONBLOCK) < 0)
	{
		std::fprintf(stderr, "%s: could not set TCP socket to non-blocking mode (psock_ctl: %s)\n", argv[0], std::strerror(psock_errno()));
		psock_close(sd);
		if (changed_in)
			tcsetattr(fd_stdin, TCSAFLUSH, &in_attr_old);
		return EXIT_FAILURE;
	}

	std::printf("Connected to %s:%hu.\n", inet_ntoa(saddr.sin_addr), servport);
#ifdef CLIENTDEBUG
	eventlog_set(stderr);
#endif

	if (!(packet = packet_create(packet_class_init)))
	{
		std::fprintf(stderr, "%s: could not create packet\n", argv[0]);
		if (changed_in)
			tcsetattr(fd_stdin, TCSAFLUSH, &in_attr_old);
		return EXIT_FAILURE;
	}
	bn_byte_set(&packet->u.client_initconn.cclass, CLIENT_INITCONN_CLASS_BOT);
	client_blocksend_packet(sd, packet);
	packet_del_ref(packet);

	if (!(rpacket = packet_create(packet_class_raw)))
	{
		std::fprintf(stderr, "%s: could not create packet\n", argv[0]);
		if (changed_in)
			tcsetattr(fd_stdin, TCSAFLUSH, &in_attr_old);
		return EXIT_FAILURE;
	}

	if (!(packet = packet_create(packet_class_raw)))
	{
		std::fprintf(stderr, "%s: could not create packet\n", argv[0]);
		if (changed_in)
			tcsetattr(fd_stdin, TCSAFLUSH, &in_attr_old);
		return EXIT_FAILURE;
	}
	packet_append_ntstring(packet, "\004");
	client_blocksend_packet(sd, packet);
	packet_del_ref(packet);

	{
		int            highest_fd;
		t_psock_fd_set rfds;

		PSOCK_FD_ZERO(&rfds);
		highest_fd = fd_stdin;
		if (sd > highest_fd)
			highest_fd = sd;

		commpos = 0;
		text[0] = '\0';

		for (;;)
		{
			PSOCK_FD_SET(fd_stdin, &rfds);
			PSOCK_FD_SET(sd, &rfds);

			if (psock_select(highest_fd + 1, &rfds, NULL, NULL, NULL) < 0)
			{
				if (errno != PSOCK_EINTR)
					std::fprintf(stderr, "%s: select failed (select: %s)\n", argv[0], std::strerror(errno));
				continue;
			}

			if (PSOCK_FD_ISSET(sd, &rfds)) /* got network data */
			{
				packet_set_size(rpacket, MAX_PACKET_SIZE - 1);
				currsize = 0;
				if (net_recv_packet(sd, rpacket, &currsize)<0)
				{
					psock_close(sd);
					sd = -1;
				}

				if (currsize>0)
				{
					char * str = (char*)packet_get_raw_data(rpacket, 0);

					str[currsize] = '\0';
					std::printf("%s", str);
					std::fflush(stdout);
				}

				if (sd == -1) /* if connection was closed */
				{
					std::printf("Connection closed by server.\n");
					if (changed_in)
						tcsetattr(fd_stdin, TCSAFLUSH, &in_attr_old);
					packet_del_ref(rpacket);
					return EXIT_SUCCESS;
				}
			}

			if (PSOCK_FD_ISSET(fd_stdin, &rfds)) /* got keyboard data */
			{
				switch (client_get_comm("", text, sizeof(text), &commpos, -1, 0, screen_width))
				{
				case -1: /* cancel */
					if (changed_in)
						tcsetattr(fd_stdin, TCSAFLUSH, &in_attr_old);
					return EXIT_FAILURE;

				case 0: /* timeout */
					break;

				case 1:
					if (!(packet = packet_create(packet_class_raw)))
					{
						std::fprintf(stderr, "%s: could not create packet\n", argv[0]);
						if (changed_in)
							tcsetattr(fd_stdin, TCSAFLUSH, &in_attr_old);
						packet_del_ref(rpacket);
						return EXIT_FAILURE;
					}
					packet_append_ntstring(packet, text);
					packet_append_ntstring(packet, "\r\n");
					client_blocksend_packet(sd, packet);
					packet_del_ref(packet);
					commpos = 0;
					text[0] = '\0';
				}
			}
		}
	}

	/* not reached */
}
